src/mds/DamageTable.cc

```diff
@@ -0,0 +1,141 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software 
+ * Foundation.  See file COPYING.
+ * 
+ */
+
+#include "common/debug.h"
+
+#include "mds/CDir.h"
+
+#include "DamageTable.h"
+
+#define dout_subsys ceph_subsys_mds
+#undef dout_prefix
+#define dout_prefix *_dout << "mds." << rank << ".damage " << __func__ << " "
+
+
+DamageEntry::~DamageEntry()
+{}
+
+bool DamageTable::notify_dentry(
+    inodeno_t ino, frag_t frag,
+    snapid_t snap_id, const std::string &dname)
+{
+  if (oversized()) {
+    return true;
+  }
+
+  // Special cases: damage to these dirfrags is considered fatal to
+  // the MDS rank that owns them.
+  if (
+      (MDS_INO_IS_MDSDIR(ino) && MDS_INO_MDSDIR_OWNER(ino) == rank)
+      ||
+      (MDS_INO_IS_STRAY(ino) && MDS_INO_STRAY_OWNER(ino) == rank)
+     ) {
+    derr << "Damage to dentries in fragment " << frag << " of ino " << ino
+         << "is fatal because it is a system directory for this rank" << dendl;
+    return true;
+  }
+
+  DamageEntryRef entry = std::make_shared<DentryDamage>(
+      ino, frag, dname, snap_id);
+  dentries[DirFragIdent(ino, frag)][DentryIdent(dname, snap_id)] = entry;
+  by_id[entry->id] = entry;
+
+  return false;
+}
+
+bool DamageTable::notify_dirfrag(inodeno_t ino, frag_t frag)
+{
+  // Special cases: damage to these dirfrags is considered fatal to
+  // the MDS rank that owns them.
+  if (/*
+      (MDS_INO_IS_MDSDIR(ino) && MDS_INO_MDSDIR_OWNER(ino) == rank)
+      ||*/
```

Comment:
Ah, this was something that was removed because MDSDIR is auto-recreated so shouldn't really be considered fatal.  Removed commented code.